File: backend/worker.ts
#!/usr/bin/env tsx
/**
 * Retry worker: periodically retries emails that are stuck in the pipeline.
 *
 * Emails end up stuck when the extract server was down at ingestion time
 * (status='ingested') or when a previous extraction attempt failed (status='failed').
 *
 * Entry point:
 *   tsx backend/worker.ts
 *
 * Env vars (all optional — defaults work for local dev):
 *   EXTRACT_SERVER_URL      default http://localhost:8001
 *   WORKER_INTERVAL_SECONDS default 60
 *   WORKER_GRACE_SECONDS    default 120  (skip emails ingested less than N seconds ago)
 *   WORKER_MAX_RETRIES      default 3    (don't retry failed emails more than N times)
 */

import { pool } from './db.js'

const EXTRACT_SERVER_URL    = process.env.EXTRACT_SERVER_URL ?? 'http://localhost:8001'
const INTERVAL_SECONDS      = Number(process.env.WORKER_INTERVAL_SECONDS ?? 60)
const GRACE_SECONDS         = Number(process.env.WORKER_GRACE_SECONDS    ?? 120)
const MAX_RETRIES           = Number(process.env.WORKER_MAX_RETRIES      ?? 3)

// ── email discovery ───────────────────────────────────────────────────────────

interface StuckEmail {
  id: string
  status: string
  sender: string
  attempt_count: number
}

async function findStuckEmails(): Promise<StuckEmail[]> {
  const { rows } = await pool.query<StuckEmail>(
    `SELECT
       e.id,
       e.status,
       e.sender,
       count(er.id)::integer AS attempt_count
     FROM emails e
     LEFT JOIN extraction_runs er ON er.email_id = e.id
     WHERE (
       -- Ingested but pipeline never started, or extract server was down
       e.status = 'ingested'
       AND e.received_at < now() - make_interval(secs => $1)
       AND NOT EXISTS (
         SELECT 1 FROM extraction_runs er2
         WHERE er2.email_id = e.id
           AND er2.created_at > now() - interval '10 minutes'
       )
     ) OR (
       -- Previously failed but still under the retry limit
       e.status = 'failed'
       AND e.received_at < now() - interval '5 minutes'
     )
     GROUP BY e.id, e.status, e.sender
     HAVING count(er.id) < $2
     LIMIT 20`,
    [GRACE_SECONDS, MAX_RETRIES],
  )
  return rows
}

// ── pipeline trigger ──────────────────────────────────────────────────────────

async function triggerPipeline(emailId: string): Promise<void> {
  const res = await fetch(`${EXTRACT_SERVER_URL}/pipeline`, {
    method: 'POST',
    body: emailId,
  })
  if (!res.ok) {
    const body = await res.text().catch(() => '')
    throw new Error(`pipeline ${res.status}: ${body}`)
  }
}

// ── worker loop ───────────────────────────────────────────────────────────────

async function tick(): Promise<void> {
  const stuck = await findStuckEmails()
  if (stuck.length === 0) return

  console.log(`[worker] ${new Date().toISOString()} — retrying ${stuck.length} stuck email(s)`)

  for (const email of stuck) {
    try {
      await triggerPipeline(email.id)
      console.log(
        `[worker]   ✓ retriggered ${email.id}` +
        `  status=${email.status}  attempts=${email.attempt_count}` +
        `  from=${email.sender}`,
      )
    } catch (err) {
      console.error(
        `[worker]   ✗ failed to retrigger ${email.id}:`,
        err instanceof Error ? err.message : err,
      )
    }
  }
}

async function run(): Promise<void> {
  console.log(
    `[worker] starting` +
    `  interval=${INTERVAL_SECONDS}s` +
    `  grace=${GRACE_SECONDS}s` +
    `  max_retries=${MAX_RETRIES}` +
    `  extract=${EXTRACT_SERVER_URL}`,
  )

  // Run once immediately on startup to catch anything stuck before the worker launched.
  await tick().catch((err) => console.error('[worker] initial tick error:', err))

  setInterval(() => {
    tick().catch((err) => console.error('[worker] tick error:', err))
  }, INTERVAL_SECONDS * 1000)
}

run().catch((err) => {
  console.error('[worker] fatal:', err)
  process.exit(1)
})

File: backend/worker.py
"""Retry worker: an async polling loop inside the same FastAPI process.

Emails get stuck when the pipeline never started (status='ingested') or when
a previous extraction errored (status='failed'). The worker finds them and
re-runs the pipeline, bounded by WORKER_MAX_RETRIES so a permanently broken
email cannot spin forever.

It runs as a background task started in the app lifespan rather than as a
separate process, so `uvicorn backend.main:app` is the whole backend. Set
WORKER_ENABLED=false to turn it off.
"""

from __future__ import annotations

import asyncio
import logging
from typing import Any

from backend.config import get_settings
from backend.db import connection
from backend.pipeline import run_pipeline

log = logging.getLogger(__name__)

# Ingested but the pipeline never started, or previously failed and still
# under the retry limit. Emails younger than the grace period are left alone
# so the worker never races the request that just created them.
_STUCK_SQL = """
SELECT
  e.id,
  e.status,
  e.sender,
  count(er.id)::integer AS attempt_count
FROM emails e
LEFT JOIN extraction_runs er ON er.email_id = e.id
WHERE (
  e.status = 'ingested'
  AND e.received_at < now() - make_interval(secs => %(grace)s)
  AND NOT EXISTS (
    SELECT 1 FROM extraction_runs er2
    WHERE er2.email_id = e.id
      AND er2.created_at > now() - interval '10 minutes'
  )
) OR (
  e.status = 'failed'
  AND e.received_at < now() - interval '5 minutes'
)
GROUP BY e.id, e.status, e.sender
HAVING count(er.id) < %(max_retries)s
LIMIT 20
"""


async def find_stuck_emails() -> list[dict[str, Any]]:
    settings = get_settings()
    async with connection() as conn:
        cur = await conn.execute(
            _STUCK_SQL,
            {
                "grace": settings.worker_grace_seconds,
                "max_retries": settings.worker_max_retries,
            },
        )
        return await cur.fetchall()


async def tick() -> int:
    """One pass. Returns how many emails were successfully re-triggered."""
    stuck = await find_stuck_emails()
    if not stuck:
        return 0

    log.info("retrying %d stuck email(s)", len(stuck))
    retriggered = 0
    for email in stuck:
        try:
            await run_pipeline(str(email["id"]))
            retriggered += 1
            log.info(
                "retriggered %s status=%s attempts=%s from=%s",
                email["id"],
                email["status"],
                email["attempt_count"],
                email["sender"],
            )
        except Exception as err:
            log.error("failed to retrigger %s: %s", email["id"], err)
    return retriggered


async def run_worker() -> None:
    """Poll forever. Cancelled by the app lifespan on shutdown."""
    settings = get_settings()
    log.info(
        "worker starting interval=%ss grace=%ss max_retries=%s",
        settings.worker_interval_seconds,
        settings.worker_grace_seconds,
        settings.worker_max_retries,
    )
    while True:
        try:
            await tick()
        except asyncio.CancelledError:
            log.info("worker stopping")
            raise
        except Exception as err:
            log.error("worker tick error: %s", err)
        await asyncio.sleep(settings.worker_interval_seconds)
